state/processor.go

```diff
@@ -94,60 +91,58 @@ func (tp *TransactionProcessor) ValidateSignature(s types.Signed) (types.Address
 	return addr, nil
 }
 
-// Validate the tx's signature by extracting the source account and validating its existence.
-// Return the src acount address and error in case of failure
-func (tp *TransactionProcessor) ValidateTransactionSignature(tx *types.SerializableSignedTransaction) (types.Address, error) {
-	buf, err := types.InterfaceToBytes(&tx.InnerSerializableSignedTransaction)
-	if err != nil {
-		return types.Address{}, err
-	}
-	pubKey, err := ed25519.ExtractPublicKey(buf, tx.Signature[:])
-	if err != nil {
-		return types.Address{}, err
-	}
-
-	addr := PublicKeyToAccountAddress(pubKey)
-	if !tp.globalState.Exist(addr) {
-		return types.Address{}, fmt.Errorf("failed to validate tx signature, unknown src account %v", addr)
-	}
-
-	return addr, nil
+// AddressExists checks if an account address exists in this node's global state
+func (tp *TransactionProcessor) AddressExists(addr types.Address) bool {
+	return tp.globalState.Exist(addr)
 }
 
-func (tp *TransactionProcessor) GetValidAddressableTx(tx *types.SerializableSignedTransaction) (*types.AddressableSignedTransaction, error) {
-	addr, err := tp.ValidateTransactionSignature(tx)
+func (tp *TransactionProcessor) ValidateNonceAndBalance(tx *types.Transaction) error {
+	origin := tx.Origin()
+	nonce, balance, err := tp.projector.GetProjection(origin, tp.globalState.GetNonce(origin), tp.globalState.GetBalance(origin))
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("failed to project state for account %v: %v", origin.Short(), err)
 	}
-
-	return &types.AddressableSignedTransaction{SerializableSignedTransaction: tx, Address: addr}, nil
+	if tx.AccountNonce != nonce {
+		return fmt.Errorf("incorrect account nonce! Expected: %d, Actual: %d", nonce, tx.AccountNonce)
+	}
+	if (tx.Amount + tx.GasPrice) > balance { // TODO: GasPrice represents the absolute fee here, as a temporarily hack
+		return fmt.Errorf("insufficient balance! Available: %d, Attempting to spend: %d[amount]+%d[fee]=%d",
+			balance, tx.Amount, tx.GasPrice, tx.Amount+tx.GasPrice)
+	}
+	return nil
 }
 
-//should receive sort predicate
 // ApplyTransaction receives a batch of transaction to apply on state. Returns the number of transaction that failed to apply.
-func (tp *TransactionProcessor) ApplyTransactions(layer types.LayerID, txs mesh.Transactions) (uint32, error) {
-	//todo: need to seed the mersenne twister with random beacon seed
+func (tp *TransactionProcessor) ApplyTransactions(layer types.LayerID, txs []*types.Transaction) (int, error) {
 	if len(txs) == 0 {
 		return 0, nil
 	}
 
-	//txs := MergeDoubles(mesh.Transactions)
 	tp.mu.Lock()
 	defer tp.mu.Unlock()
-	failed := tp.Process(tp.randomSort(txs), tp.coalescTransactionsBySender(txs))
+	remaining := txs
+	remainingCount := len(remaining)
+	for { // Loop until there's nothing left to process
+		remaining = tp.Process(remaining)
```

Comment:
if a transaction is never processed this loop will be infinite

Reply:
The loops stop condition isn't that `len(remaining)` is zero, it's that it's the same as before the pass. If all processable txs have been processed, `len(remaining)` will equal its previous value and the loop will break.